memory_reference/src/algo.hpp

```diff
@@ -109,4 +123,65 @@ class Algo<CUDAGPU> {
         cudaMemcpy(output.ptr, input.ptr, size * sizeof(double), cudaMemcpyDeviceToDevice);
     }
 };
-#endif 
\ No newline at end of file
+#endif
+
+#ifdef ENABLED_HIP
+template<>
+class Algo<HIPGPU> {
+public:
+
+    void whoami()
+    {
+        std::cout << "I'm the HIP GPU version\n";
+    }
+
+    void compute(Mem_Ref<HIPGPU> input, Mem_Ref<HIPGPU> output, int size)
+    {
+        hipMemcpy(output.ptr, input.ptr, size * sizeof(double), hipMemcpyDeviceToDevice);
+    }
+};
+
+class Hip_Mem_Impl {
+public:
+
+    int size;
+    Hip_Mem_Impl(){}
+
+    Mem_Ref<HIPGPU> get_hip_array(int size) 
+    {
+        this->size = size;
+        Mem_Ref<HIPGPU> hip_array;
+
+        std::vector<double> cpu_array;
+        for(int i = 0; i < size; i++) cpu_array.push_back(0.0);
```

Comment:
There is a more easy way to create an vector with 0's: `std::vector<int> vector1(length, 0);`
https://stackoverflow.com/questions/13110130/initialize-a-vector-to-zeros-c-c11

By the way, the `push_back` approach is pretty inefficient, because it trigger relocating of memory.